markov_builder/MarkovChain.py

```
@@ -0,0 +1,878 @@
+from typing import List, Optional, Tuple
+
+import myokit
+import networkx as nx
+import numpy as np
+import pandas as pd
+import pyvis
+import sympy as sp
+from numpy.random import default_rng
+
+
+@dataclass
+class MarkovStateAttributes:
+    """A dataclass defining what attributes each state in a MarkovChain should have, and their default values"""
+    open_state: bool = False
+    inactive: bool = False
+
+
+class MarkovChain():
+    """A class used to construct continuous time Markov Chains/compartmental models using networkx.
+
+    Various helper functions are included to generate equations, code for the
+    Markov chains and to test whether the model has certain properties.
+    """
+
+    def __init__(self, states: Optional[list] = None, state_attributes_class:
+                 Optional[MarkovStateAttributes] = None, seed: Optional[int] = None, name:
+                 Optional[str] = None):
+
+        # Initialise the graph representing the states. Each directed edge has
+        # a `rate` attribute which is a string representing the transition rate
+        # between the two nodes
+
+        self.graph = nx.DiGraph()
+        if states is not None:
+            self.graph.add_nodes_from(states)
+        self.rates = set()
+
+        # Initialise a random number generator for simulation. Optionally, a
+        # seed can be specified.
+        self.rng = default_rng(seed)
+        self.name = name
+        self.shared_rate_variables = set()
+
+        self.rate_expressions = {}
+        self.default_values = {}
+
+        self.auxiliary_expression = None
+
+        if state_attributes_class is None:
+            state_attributes_class = MarkovStateAttributes
+
+        self.state_attributes_class = state_attributes_class
+        self.reserved_names = []
+        self.auxiliary_variable = 'auxiliary_expression'
+
+        if not is_dataclass(self.state_attributes_class):
+            raise Exception("state_attirbutes_class must be a dataclass")
```

If you can always better to raise a specific exception that indicates what the error is. 
A good chocie here would probably be `TypeError`